homeassistant/helpers/script.py

```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
+        self._last_action = value
 
-        self._referenced_entities = referenced
-        return referenced
-
-    def run(self, variables=None, context=None):
+    @abstractmethod
+    async def async_run(self) -> None:
         """Run script."""
-        asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
-        ).result()
-
-    async def async_run(
-        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
 
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
+    async def _async_run(self, start):
+        for self._step, self._action in islice(
+            enumerate(self._script.sequence), start, None
+        ):
             try:
-                await self._handle_action(action, variables, context)
+                await getattr(self, f"_async_{_determine_action(self._action)}_step")()
             except _SuspendScript:
-                # Store next step to take and notify change listeners
-                self._cur = cur + 1
-                if self._change_listener:
-                    self.hass.async_add_job(self._change_listener)
-                return
+                raise
             except _StopScript:
                 break
-            except Exception:
-                # Store the step that had an exception
-                self._exception_step = cur
-                # Set script to not running
-                self._cur = -1
-                self.last_action = None
+            except Exception as err:
+                self._log_exception(err)
+                self._stop()
                 # Pass exception on.
                 raise
 
-        # Set script to not-running.
-        self._cur = -1
-        self.last_action = None
+        self._stop()
         if self._change_listener:
-            self.hass.async_add_job(self._change_listener)
+            self._hass.async_add_job(self._change_listener)
 
+    @abstractmethod
     def stop(self) -> None:
-        """Stop running script."""
-        run_callback_threadsafe(self.hass.loop, self.async_stop).result()
-
-    @callback
-    def async_stop(self) -> None:
-        """Stop running script."""
-        if self._cur == -1:
-            return
+        """Stop script run."""
 
-        self._cur = -1
-        self._async_remove_listener()
-        if self._change_listener:
-            self.hass.async_add_job(self._change_listener)
+    def _stop(self):
+        self._script._runs.remove(self)  # pylint: disable=protected-access
```

Comment:
Why have this? Shouldn't we just have the script register itself as a change_listener?

Reply:
This code evolved quite a bit as I struggled to reorganize / clean things up, while at the same time still being able to support the "legacy" behavior. I found this the cleanest way. Besides, the change_listener is one function/method passed into Script. I didn't extend the concept by allowing multiple change listeners, or to somehow use that mechanism between Script and _ScriptRunXXX. Bottom line, when a script run completes, or is told to stop, it needs to be removed from the list of "runs", and I found this the cleanest way to do it.